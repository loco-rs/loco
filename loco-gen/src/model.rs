use std::{collections::HashMap, env::current_dir};

use chrono::Utc;
use duct::cmd;
use rrgen::RRgen;
use serde_json::json;

use super::{Error, Result};
use crate::get_mappings;

const MODEL_T: &str = include_str!("templates/model/model.t");
const MODEL_TEST_T: &str = include_str!("templates/model/test.t");

use super::{collect_messages, AppInfo};

/// skipping some fields from the generated models.
/// For example, the `created_at` and `updated_at` fields are automatically
/// generated by the Loco app and should be given
pub const IGNORE_FIELDS: &[&str] = &["created_at", "updated_at", "create_at", "update_at"];

pub fn generate(
    rrgen: &RRgen,
    name: &str,
    is_link: bool,
    migration_only: bool,
    fields: &[(String, String)],
    appinfo: &AppInfo,
) -> Result<String> {
    let pkg_name: &str = &appinfo.app_name;
    let ts = Utc::now();

    let mut columns = Vec::new();
    let mut references = Vec::new();
    for (fname, ftype) in fields {
        if IGNORE_FIELDS.contains(&fname.as_str()) {
            tracing::warn!(
                field = fname,
                "note that a redundant field was specified, it is already generated automatically"
            );
            continue;
        }
        if ftype == "references" {
            let fkey = format!("{fname}_id");
            columns.push((fkey.clone(), "integer"));
            // user, user_id
            references.push((fname.to_string(), fkey));
        } else if let Some(refname) = ftype.strip_prefix("references:") {
            let fkey = format!("{fname}_id");
            columns.push((fkey.clone(), "integer"));
            references.push((refname.to_string(), fkey));
        } else {
            let mappings = get_mappings();
            let schema_type = mappings.schema_field(ftype.as_str()).ok_or_else(|| {
                Error::Message(format!(
                    "type: {} not found. try any of: {:?}",
                    ftype,
                    mappings.schema_fields()
                ))
            })?;
            columns.push((fname.to_string(), schema_type.as_str()));
        }
    }

    let vars = json!({"name": name, "ts": ts, "pkg_name": pkg_name, "is_link": is_link, "columns": columns, "references": references});
    let res1 = rrgen.generate(MODEL_T, &vars)?;
    let res2 = rrgen.generate(MODEL_TEST_T, &vars)?;

    if !migration_only {
        let cwd = current_dir()?;
        let env_map: HashMap<_, _> = std::env::vars().collect();

        let _ = cmd!("cargo", "loco-tool", "db", "migrate",)
            .stderr_to_stdout()
            .dir(cwd.as_path())
            .full_env(&env_map)
            .run()
            .map_err(|err| {
                Error::Message(format!(
                    "failed to run loco db migration. error details: `{err}`",
                ))
            })?;
        let _ = cmd!("cargo", "loco-tool", "db", "entities",)
            .stderr_to_stdout()
            .dir(cwd.as_path())
            .full_env(&env_map)
            .run()
            .map_err(|err| {
                Error::Message(format!(
                    "failed to run loco db entities. error details: `{err}`",
                ))
            })?;
    }

    let messages = collect_messages(vec![res1, res2]);
    Ok(messages)
}

#[cfg(test)]
mod tests {
    use std::{env, process::Command};

    use crate::{
        testutil::{self, assert_cargo_check, assert_file, assert_single_file_match},
        AppInfo,
    };

    fn with_new_app<F>(app_name: &str, f: F)
    where
        F: FnOnce(),
    {
        testutil::with_temp_dir(|_previous, current| {
            let status = Command::new("loco")
                .args([
                    "new",
                    "-n",
                    app_name,
                    "--db",
                    "sqlite",
                    "--bg",
                    "async",
                    "--assets",
                    "serverside",
                    "-a",
                ])
                .status()
                .expect("cannot run command");

            assert!(status.success(), "Command failed: loco new -n {app_name}");
            env::set_current_dir(current.join(app_name))
                .expect("Failed to change directory to app");
            f(); // Execute the provided closure
        })
        .expect("temp dir setup");
    }

    #[test]
    fn test_can_generate_model() {
        let rrgen = rrgen::RRgen::default();
        with_new_app("saas", || {
            super::generate(
                &rrgen,
                "movies",
                false,
                true,
                &[("title".to_string(), "string".to_string())],
                &AppInfo {
                    app_name: "saas".to_string(),
                },
            )
            .expect("generate");
            assert_file("migration/src/lib.rs", |content| {
                content.assert_syntax();
                content.assert_regex_match("_movies::Migration");
            });
            let migration = assert_single_file_match("migration/src", ".*_movies.rs$");
            assert_file(migration.to_str().unwrap(), |content| {
                content.assert_syntax();
                content.assert_regex_match("Title");
            });
            assert_cargo_check();
        });
    }
}
