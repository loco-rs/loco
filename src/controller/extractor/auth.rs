//! Axum middleware for validating token header
//!
//! # Example:
//!
//! ```
//! use loco_rs::prelude::*;
//! use serde::Serialize;
//! use axum::extract::State;
//! use loco_rs::controller::extractor::auth;
//!
//! #[derive(Serialize)]
//! pub struct TestResponse {
//!     pub pid: String,
//! }
//!
//! async fn current(
//!     auth: auth::JWT,
//!     State(ctx): State<AppContext>,
//! ) -> Result<Response> {
//!     format::json(TestResponse{ pid: auth.claims.pid})
//! }
//! ```
use std::collections::HashMap;

use axum::{
    extract::{FromRef, FromRequestParts, OptionalFromRequestParts, Query},
    http::{request::Parts, HeaderMap},
};
use axum_extra::extract::cookie;
use serde::{Deserialize, Serialize};
use tracing;

use crate::{app::AppContext, auth, config::JWT as JWTConfig, errors::Error, Result as LocoResult};

#[cfg(feature = "with-db")]
use crate::model::{Authenticable, ModelError};

// ---------------------------------------
//
// JWT Auth extractor
//
// ---------------------------------------

// Define constants for token prefix and authorization header
const TOKEN_PREFIX: &str = "Bearer ";
const AUTH_HEADER: &str = "authorization";

// Define a struct to represent user authentication information serialized
// to/from JSON
#[cfg(feature = "with-db")]
#[derive(Debug, Deserialize, Serialize)]
pub struct JWTWithUser<T: Authenticable> {
    pub claims: auth::jwt::UserClaims,
    pub user: T,
}

// Implement axum traits for the Auth struct
#[cfg(feature = "with-db")]
impl<S, T> FromRequestParts<S> for JWTWithUser<T>
where
    AppContext: FromRef<S>,
    S: Send + Sync,
    T: Authenticable,
{
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Error> {
        let ctx: AppContext = AppContext::from_ref(state);

        let token = extract_token(get_jwt_from_config(&ctx)?, parts)?.ok_or_else(
            // If we get here, none of the locations worked
            || Error::Unauthorized("Token not found in any of the configured JWT locations. Please check your auth.jwt.location configuration.".to_string()))?;

        let jwt_secret = ctx.config.get_jwt_config()?;

        match auth::jwt::JWT::new(&jwt_secret.secret).validate(&token) {
            Ok(claims) => {
                let user = T::find_by_claims_key(&ctx.db, &claims.claims.pid)
                    .await
                    .map_err(|e| match e {
                        ModelError::EntityNotFound => Error::Unauthorized("not found".to_string()),
                        ModelError::DbErr(db_err) => {
                            tracing::error!("Database error during authentication: {}", db_err);
                            Error::InternalServerError
                        }
                        _ => {
                            tracing::error!("Authentication error: {}", e);
                            Error::Unauthorized("could not authorize".to_string())
                        }
                    })?;
                Ok(Self {
                    claims: claims.claims,
                    user,
                })
            }
            Err(err) => {
                tracing::error!("JWT validation error: {}", err);
                Err(Error::Unauthorized("token is not valid".to_string()))
            }
        }
    }
}

#[cfg(feature = "with-db")]
impl<S, T> OptionalFromRequestParts<S> for JWTWithUser<T>
where
    AppContext: FromRef<S>,
    S: Send + Sync,
    T: Authenticable,
{
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Option<Self>, Error> {
        let ctx: AppContext = AppContext::from_ref(state);

        if let Some(token) = extract_token(get_jwt_from_config(&ctx)?, parts)? {
            let jwt_secret = ctx.config.get_jwt_config()?;

            match auth::jwt::JWT::new(&jwt_secret.secret).validate(&token) {
                Ok(claims) => match T::find_by_claims_key(&ctx.db, &claims.claims.pid).await {
                    Ok(user) => Ok(Some(JWTWithUser {
                        claims: claims.claims,
                        user,
                    })),
                    Err(ModelError::EntityNotFound) => {
                        Err(Error::Unauthorized("not found".to_string()))
                    }
                    Err(ModelError::DbErr(db_err)) => {
                        tracing::error!("Database error during authentication: {}", db_err);
                        Err(Error::InternalServerError)
                    }
                    Err(unhandled) => {
                        tracing::error!("Authentication error: {}", unhandled);
                        Err(Error::Unauthorized("could not authorize".to_string()))
                    }
                },
                Err(err) => {
                    tracing::error!("JWT validation error: {}", err);
                    Err(Error::Unauthorized("token is not valid".to_string()))
                }
            }
        } else {
            Ok(None)
        }
    }
}

// Define a struct to represent user authentication information serialized
// to/from JSON
#[derive(Debug, Deserialize, Serialize)]
pub struct JWT {
    pub claims: auth::jwt::UserClaims,
}

// Implement the FromRequestParts trait for the Auth struct
impl<S> FromRequestParts<S> for JWT
where
    AppContext: FromRef<S>,
    S: Send + Sync,
{
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Error> {
        extract_jwt_from_request_parts(parts, state)?.ok_or_else(
            // If we get here, none of the locations worked
            || Error::Unauthorized("Token not found in any of the configured JWT locations. Please check your auth.jwt.location configuration.".to_string()))
    }
}

// Implement the OptionalFromRequestParts trait for the Auth struct, for routes where behaviour depends on auth status.
impl<S> OptionalFromRequestParts<S> for JWT
where
    AppContext: FromRef<S>,
    S: Send + Sync,
{
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Option<Self>, Error> {
        extract_jwt_from_request_parts(parts, state)
    }
}

/// extract a [JWT] token from request parts, using a non-mutable reference to the [Parts]
///
/// # Errors
/// Return an error when JWT token not configured or when the token is not valid
pub fn extract_jwt_from_request_parts<S>(parts: &Parts, state: &S) -> Result<Option<JWT>, Error>
where
    AppContext: FromRef<S>,
    S: Send + Sync,
{
    let ctx: AppContext = AppContext::from_ref(state); // change to ctx

    if let Some(token) = extract_token(get_jwt_from_config(&ctx)?, parts)? {
        let jwt_secret = ctx.config.get_jwt_config()?;

        match auth::jwt::JWT::new(&jwt_secret.secret).validate(&token) {
            Ok(claims) => Ok(Some(JWT {
                claims: claims.claims,
            })),
            Err(err) => {
                tracing::error!("JWT validation error: {}", err);
                Err(Error::Unauthorized("token is not valid".to_string()))
            }
        }
    } else {
        Ok(None)
    }
}

/// extract JWT token from context configuration
///
/// # Errors
/// Return an error when JWT token not configured
pub fn get_jwt_from_config(ctx: &AppContext) -> LocoResult<&JWTConfig> {
    ctx.config
        .auth
        .as_ref()
        .ok_or_else(|| Error::string("auth not configured"))?
        .jwt
        .as_ref()
        .ok_or_else(|| Error::string("JWT token not configured"))
}
/// extract token from the configured jwt location settings
///
/// # Errors
///
/// Returns an error when the token cannot be extracted from any of the configured locations,
/// such as missing headers, invalid formats, or inaccessible request data.
pub fn extract_token(jwt_config: &JWTConfig, parts: &Parts) -> LocoResult<Option<String>> {
    let locations = get_jwt_locations(jwt_config.location.as_ref());

    for location in locations {
        if let Ok(Some(token)) = extract_token_from_location(location, parts) {
            return Ok(Some(token));
        }
    }

    Ok(None)
}

/// Get the list of JWT locations to try, with Bearer as default
fn get_jwt_locations(
    config: Option<&crate::config::JWTLocationConfig>,
) -> Vec<&crate::config::JWTLocation> {
    match config {
        Some(crate::config::JWTLocationConfig::Single(location)) => vec![location],
        Some(crate::config::JWTLocationConfig::Multiple(locations)) => locations.iter().collect(),
        None => vec![&crate::config::JWTLocation::Bearer],
    }
}

/// Extract token from a specific location
fn extract_token_from_location(
    location: &crate::config::JWTLocation,
    parts: &Parts,
) -> LocoResult<Option<String>> {
    match location {
        crate::config::JWTLocation::Query { name } => extract_token_from_query(name, parts),
        crate::config::JWTLocation::Cookie { name } => extract_token_from_cookie(name, parts),
        crate::config::JWTLocation::Bearer => extract_token_from_header(&parts.headers),
    }
}

/// Function to extract a token from the authorization header
///
/// # Errors
///
/// When token is not valid or not found
pub fn extract_token_from_header(headers: &HeaderMap) -> LocoResult<Option<String>> {
    if let Some(token) = headers.get(AUTH_HEADER) {
        Ok(Some(
            token
                .to_str()
                .map_err(|err| Error::Unauthorized(err.to_string()))?
                .strip_prefix(TOKEN_PREFIX)
                .ok_or_else(|| Error::Unauthorized(format!("error strip {AUTH_HEADER} value")))?
                .to_string(),
        ))
    } else {
        Ok(None)
    }
}

/// Extract a token value from cookie
///
/// # Errors
/// when token value from cookie is not found
pub fn extract_token_from_cookie(name: &str, parts: &Parts) -> LocoResult<Option<String>> {
    // LogoResult
    let jar: cookie::CookieJar = cookie::CookieJar::from_headers(&parts.headers);
    Ok(jar.get(name).map(|c| c.value().to_string()))
}
/// Extract a token value from query
///
/// # Errors
/// when token value from cookie is not found
pub fn extract_token_from_query(name: &str, parts: &Parts) -> LocoResult<Option<String>> {
    // LogoResult
    if let Ok(parameters) = Query::<HashMap<String, String>>::try_from_uri(&parts.uri) {
        Ok(parameters.get(name).cloned())
    } else {
        Ok(None)
    }
}

// ---------------------------------------
//
// API Token Auth / Extractor
//
// ---------------------------------------
#[cfg(feature = "with-db")]
#[derive(Debug, Deserialize, Serialize)]
// Represents the data structure for the API token.
pub struct ApiToken<T: Authenticable> {
    pub user: T,
}

// Implementing the `FromRequestParts` trait for `ApiToken` to enable extracting
// it from the request.
#[cfg(feature = "with-db")]
impl<S, T> FromRequestParts<S> for ApiToken<T>
where
    AppContext: FromRef<S>,
    S: Send + Sync,
    T: Authenticable,
{
    type Rejection = Error;

    // Extracts `ApiToken` from the request parts.
    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Error> {
        // Extract API key from the request header.
        let api_key = extract_token_from_header(&parts.headers)?
            .ok_or_else(|| Error::Unauthorized(format!("header {AUTH_HEADER} token not found")))?;

        // Convert the state reference to the application context.
        let state: AppContext = AppContext::from_ref(state);

        // Retrieve user information based on the API key from the database.
        let user = T::find_by_api_key(&state.db, &api_key)
            .await
            .map_err(|e| match e {
                ModelError::EntityNotFound => Error::Unauthorized("not found".to_string()),
                ModelError::DbErr(db_err) => {
                    tracing::error!("Database error during API key authentication: {}", db_err);
                    Error::InternalServerError
                }
                _ => {
                    tracing::error!("API key authentication error: {}", e);
                    Error::Unauthorized("could not authorize".to_string())
                }
            })?;

        Ok(Self { user })
    }
}

#[cfg(test)]
mod tests {

    use axum::http::{HeaderMap, HeaderValue};

    use super::*;
    use crate::config;

    #[test]
    fn test_extract_token_from_header_success() {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTH_HEADER,
            HeaderValue::from_str("Bearer valid_token_123").unwrap(),
        );

        let result = extract_token_from_header(&headers);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("valid_token_123".to_string()));
    }

    #[test]
    fn test_extract_token_from_header_with_spaces() {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTH_HEADER,
            HeaderValue::from_str("Bearer  token_with_spaces  ").unwrap(),
        );

        let result = extract_token_from_header(&headers);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some(" token_with_spaces  ".to_string()));
    }

    #[test]
    fn test_extract_token_from_header_special_chars() {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTH_HEADER,
            HeaderValue::from_str("Bearer token-with_special.chars").unwrap(),
        );

        let result = extract_token_from_header(&headers);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("token-with_special.chars".to_string()));
    }

    #[test]
    fn test_extract_token_from_header_missing_header() {
        let headers = HeaderMap::new();
        let result = extract_token_from_header(&headers);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[test]
    fn test_extract_token_from_header_missing_bearer_prefix() {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTH_HEADER,
            HeaderValue::from_str("InvalidPrefix token").unwrap(),
        );

        let result = extract_token_from_header(&headers);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("error strip authorization value"));
    }

    #[test]
    fn test_extract_token_from_header_empty_value() {
        let mut headers = HeaderMap::new();
        headers.insert(AUTH_HEADER, HeaderValue::from_str("Bearer").unwrap());

        let result = extract_token_from_header(&headers);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("error strip authorization value"));
    }

    #[test]
    fn test_extract_token_from_cookie_success() {
        let request = axum::http::Request::builder()
            .header("Cookie", "test_cookie=cookie_value_123")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_cookie("test_cookie", &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("cookie_value_123".to_string()));
    }

    #[test]
    fn test_extract_token_from_cookie_special_chars() {
        let request = axum::http::Request::builder()
            .header("Cookie", "auth_token=token-with.special_chars")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_cookie("auth_token", &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("token-with.special_chars".to_string()));
    }

    #[test]
    fn test_extract_token_from_cookie_missing_cookie() {
        let request = axum::http::Request::builder().body(()).unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_cookie("nonexistent", &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[test]
    fn test_extract_token_from_cookie_not_found() {
        let request = axum::http::Request::builder()
            .header("Cookie", "different_cookie=value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_cookie("nonexistent", &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[test]
    fn test_extract_token_from_query_success() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?token=query_value_123&other=param")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_query("token", &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("query_value_123".to_string()));
    }

    #[test]
    fn test_extract_token_from_query_special_chars() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?auth_token=token-with.special_chars")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_query("auth_token", &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("token-with.special_chars".to_string()));
    }

    #[test]
    fn test_extract_token_from_query_missing_param() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?other=param")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_query("nonexistent_param", &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[test]
    fn test_extract_token_from_query_invalid_uri() {
        let request = axum::http::Request::builder()
            .uri("not-a-valid-uri")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_query("nonexistent_param", &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none())
    }

    #[test]
    fn test_get_jwt_locations_default() {
        let jwt_config = JWTConfig {
            location: None,
            secret: String::new(),
            expiration: 1,
        };

        let locations = get_jwt_locations(jwt_config.location.as_ref());
        assert_eq!(locations.len(), 1);
        assert!(matches!(locations[0], config::JWTLocation::Bearer));
    }

    #[test]
    fn test_get_jwt_locations_single_bearer() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Bearer,
            )),
            secret: String::new(),
            expiration: 1,
        };

        let locations = get_jwt_locations(jwt_config.location.as_ref());
        assert_eq!(locations.len(), 1);
        assert!(matches!(locations[0], config::JWTLocation::Bearer));
    }

    #[test]
    fn test_get_jwt_locations_single_cookie() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Cookie {
                    name: "auth_token".to_string(),
                },
            )),
            secret: String::new(),
            expiration: 1,
        };

        let locations = get_jwt_locations(jwt_config.location.as_ref());
        assert_eq!(locations.len(), 1);
        assert!(matches!(locations[0], config::JWTLocation::Cookie { .. }));
    }

    #[test]
    fn test_get_jwt_locations_single_query() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Query {
                    name: "token".to_string(),
                },
            )),
            secret: String::new(),
            expiration: 1,
        };

        let locations = get_jwt_locations(jwt_config.location.as_ref());
        assert_eq!(locations.len(), 1);
        assert!(matches!(locations[0], config::JWTLocation::Query { .. }));
    }

    #[test]
    fn test_get_jwt_locations_multiple() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Multiple(vec![
                config::JWTLocation::Cookie {
                    name: "auth".to_string(),
                },
                config::JWTLocation::Query {
                    name: "token".to_string(),
                },
                config::JWTLocation::Bearer,
            ])),
            secret: String::new(),
            expiration: 1,
        };

        let locations = get_jwt_locations(jwt_config.location.as_ref());
        assert_eq!(locations.len(), 3);
        assert!(matches!(locations[0], config::JWTLocation::Cookie { .. }));
        assert!(matches!(locations[1], config::JWTLocation::Query { .. }));
        assert!(matches!(locations[2], config::JWTLocation::Bearer));
    }

    #[test]
    fn test_extract_token_from_location_bearer() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?token=query_value")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_value"))
            .header("Cookie", "auth_token=cookie_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_location(&config::JWTLocation::Bearer, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some(" bearer_value".to_string()));
    }

    #[test]
    fn test_extract_token_from_location_cookie() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?token=query_value")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_value"))
            .header("Cookie", "auth_token=cookie_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_location(
            &config::JWTLocation::Cookie {
                name: "auth_token".to_string(),
            },
            &parts,
        );
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("cookie_value".to_string()));
    }

    #[test]
    fn test_extract_token_from_location_query() {
        let request = axum::http::Request::builder()
            .uri("https://example.com?token=query_value")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_value"))
            .header("Cookie", "auth_token=cookie_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token_from_location(
            &config::JWTLocation::Query {
                name: "token".to_string(),
            },
            &parts,
        );
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("query_value".to_string()));
    }

    #[test]
    fn test_extract_token_single_location_success() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Bearer,
            )),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://example.com")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} valid_token"))
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some(" valid_token".to_string()));
    }

    #[test]
    fn test_extract_token_multiple_locations_fallback() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Multiple(vec![
                config::JWTLocation::Cookie {
                    name: "nonexistent".to_string(),
                },
                config::JWTLocation::Query {
                    name: "token".to_string(),
                },
            ])),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://example.com?token=fallback_token")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("fallback_token".to_string()));
    }

    #[test]
    fn test_extract_token_all_locations_fail() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Multiple(vec![
                config::JWTLocation::Cookie {
                    name: "nonexistent".to_string(),
                },
                config::JWTLocation::Query {
                    name: "missing".to_string(),
                },
            ])),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://example.com")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[test]
    fn test_extract_from_default() {
        let jwt_config = JWTConfig {
            location: None,
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://loco.rs")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_token_value"))
            .header("Cookie", "loco_cookie_key=cookie_token_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some(" bearer_token_value".to_string()));
    }

    #[test]
    fn test_extract_from_bearer() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Bearer,
            )),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("loco.rs")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_token_value"))
            .header("Cookie", "loco_cookie_key=cookie_token_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some(" bearer_token_value".to_string()));
    }

    #[test]
    fn test_extract_from_cookie() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Cookie {
                    name: "loco_cookie_key".to_string(),
                },
            )),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://loco.rs")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_token_value"))
            .header("Cookie", "loco_cookie_key=cookie_token_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("cookie_token_value".to_string()));
    }

    #[test]
    fn test_extract_from_query() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Single(
                config::JWTLocation::Query {
                    name: "query_token".to_string(),
                },
            )),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://loco.rs?query_token=query_token_value&test=loco")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_token_value"))
            .header("Cookie", "loco_cookie_key=cookie_token_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("query_token_value".to_string()));
    }

    #[test]
    fn test_extract_from_multiple_locations() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Multiple(vec![
                config::JWTLocation::Cookie {
                    name: "nonexistent".to_string(),
                },
                config::JWTLocation::Query {
                    name: "query_token".to_string(),
                },
            ])),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://loco.rs?query_token=query_token_value&test=loco")
            .header(AUTH_HEADER, format!("{TOKEN_PREFIX} bearer_token_value"))
            .header("Cookie", "loco_cookie_key=cookie_token_value")
            .body(())
            .unwrap();
        let (parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), Some("query_token_value".to_string()));
    }

    #[test]
    fn test_extract_token_error_message_for_missing_token() {
        let jwt_config = JWTConfig {
            location: Some(config::JWTLocationConfig::Multiple(vec![
                config::JWTLocation::Cookie {
                    name: "nonexistent".to_string(),
                },
                config::JWTLocation::Query {
                    name: "missing".to_string(),
                },
            ])),
            secret: String::new(),
            expiration: 1,
        };

        let request = axum::http::Request::builder()
            .uri("https://loco.rs")
            .body(())
            .unwrap();
        let (mut parts, ()) = request.into_parts();

        let result = extract_token(&jwt_config, &parts);
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }
}
