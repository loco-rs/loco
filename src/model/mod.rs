//! # Model Error Handling
//!
//! Useful when using `sea_orm` and want to propagate errors

pub mod query;
use async_trait::async_trait;
use sea_orm::DatabaseConnection;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize)]
#[allow(clippy::module_name_repetitions)]
pub struct ModelValidation {
    pub code: String,
    pub message: Option<String>,
}

#[derive(thiserror::Error, Debug)]
#[allow(clippy::module_name_repetitions)]
pub enum ModelError {
    #[error("Entity already exists")]
    EntityAlreadyExists,

    #[error("Entity not found")]
    EntityNotFound,

    #[error("{errors:?}")]
    ModelValidation { errors: ModelValidation },

    #[cfg(feature = "auth_jwt")]
    #[error("jwt error")]
    Jwt(#[from] jsonwebtoken::errors::Error),

    #[error(transparent)]
    DbErr(#[from] sea_orm::DbErr),

    #[error(transparent)]
    Any(#[from] Box<dyn std::error::Error + Send + Sync>),

    #[error("{0}")]
    Message(String),
}

#[allow(clippy::module_name_repetitions)]
pub type ModelResult<T, E = ModelError> = std::result::Result<T, E>;

impl ModelError {
    #[must_use]
    pub fn wrap(err: impl std::error::Error + Send + Sync + 'static) -> Self {
        Self::Any(Box::new(err))
    }

    #[must_use]
    pub fn to_msg(err: impl std::error::Error + Send + Sync + 'static) -> Self {
        Self::Message(err.to_string())
    }

    #[must_use]
    pub fn msg(s: &str) -> Self {
        Self::Message(s.to_string())
    }
}
#[async_trait]
pub trait Authenticable: Clone {
    async fn find_by_api_key(db: &DatabaseConnection, api_key: &str) -> ModelResult<Self>;
    async fn find_by_claims_key(db: &DatabaseConnection, claims_key: &str) -> ModelResult<Self>;
}
