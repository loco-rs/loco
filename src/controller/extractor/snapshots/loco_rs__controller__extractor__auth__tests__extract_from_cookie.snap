---
source: src/controller/extractor/auth.rs
expression: "extract_token(&jwt_config, &parts)"
snapshot_kind: text
---
Ok(
    "cookie_token_value",
)
